dimod/views.py

```diff
@@ -72,3 +74,239 @@ class IndexValuesView(abc.ValuesView):
+
+    __slots__ = ()
+
+    def __iter__(self):
+        for v, neighbours in self._mapping._adj.items():
+            yield v, neighbours[v].value
+
+
+class QuadraticView(abc.MutableMapping):
+    """Acts as a dictionary `{(u, v): bias, ...}` for the quadratic biases.
+
+    The quadratic biases are stored in a dict-of-dicts format. So `{(u, v): bias}` is stored as
+    `._adj = {u: {v: Bias(bias)}, v: {u: Bias(bias)}}`.
+
+    """
+
+    __slots__ = '_adj',
+
+    def __init__(self, bqm):
+        self._adj = bqm._adj
+
+    def __delitem__(self, interaction):
+        u, v = interaction
+        if u == v:
+            raise KeyError('{} is not an interaction'.format(interaction))
+        adj = self._adj
+        del adj[v][u]
+        del adj[u][v]
+
+    def __getitem__(self, interaction):
+        u, v = interaction
+        if u == v:
+            raise KeyError
+        return self._adj[u][v].value
+
+    def __iter__(self):
+        seen = set()
+        adj = self._adj
+        for u, neigh in adj.items():
+
+            for v in neigh:
+                if u == v:
+                    # not adjacent to itself
+                    continue
+
+                if v not in seen:
+                    yield (u, v)
+
+            seen.add(u)
+
+    def __len__(self):
+        # -1 comes from self loops
+        return sum(len(neighbours) - 1 for neighbours in self._adj.values()) // 2
+
+    def __setitem__(self, interaction, bias):
+        u, v = interaction
+        if u == v:
+            raise KeyError('{} cannot have an interaction with itself'.format(u))
+
+        adj = self._adj
+        
+        # we don't know what type we want the biases, so we require that the variables already
+        # exist before we can add an interaction between them
+        if u not in adj:
+            raise KeyError('{} is not already a variables in the binary quadratic model'.format(u))
+        if v not in adj:
+            raise KeyError('{} is not already a variables in the binary quadratic model'.format(v))
+
+        if v in adj[u]:
+            # adj[u][v] is adj[v][u]
+            adj[u][v].value = bias
+        else:
+            adj[u][v] = adj[v][u] = Bias(bias)
+
+    def __str__(self):
+        return str(dict(self))
+
+    def items(self):
+        return QuadraticItemsView(self)
+
+
+class QuadraticItemsView(abc.ItemsView):
+    """Faster items iteration"""
+    __slots__ = ()
+
+    def __iter__(self):
+        adj = self._mapping._adj
+        for u, v in self._mapping:
+            yield (u, v), adj[u][v].value
+
+class NeighbourView(abc.Mapping):
+    """Acts as a dictionary `{u: bias, ...}` for the neightbours of a variables `v`.
+
+    See Also:
+        :class:`AdjacencyView`
+
+    """
+
+    __slots__ = '_adj', '_var'
+
+    def __init__(self, adj, v):
+        self._adj = adj
+        self._var = v
+
+    def __getitem__(self, v):
+        u = self._var
+        if u == v:
+            raise KeyError
```

Comment:
Perhaps there could be a message with the KeyError? Like `"{} is not a neighbour of itself".format(u)`